# Dataset: AB_NYC_2019

### **Author: Aregbesola Emmanuel**
### **Date: March 2026**

=====================================================================================================

In [1]:
# import pandas library

import pandas as pd

In [2]:
# load the data into pandas

df = pd.read_csv('AB_NYC_2019.csv')
pd.set_option('display.max_rows', None)

# 1. Loading and Initial Inspection

In this section, we are going to do basic inspection on our dataset. This is to get a feel of what our dataset looks like and how we are going to approach the cleaning.

We are going to check for:
1. Shape.
2. Column Names.
3. Data Types (info()).
4. Summary Statistics (describe()).
5. Duplicated Rows.
6. Null values.
7. Null Percentage Per Column.

=====================================================================================================

In [3]:
# check the number of rows and columns
df.shape

(48895, 16)

In [4]:
# check the column names for correct formatting

df.columns.to_list()

['id',
 'name',
 'host_id',
 'host_name',
 'neighbourhood_group',
 'neighbourhood',
 'latitude',
 'longitude',
 'room_type',
 'price',
 'minimum_nights',
 'number_of_reviews',
 'last_review',
 'reviews_per_month',
 'calculated_host_listings_count',
 'availability_365']

In [5]:
# check the data type of each columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

In [6]:
# summary statistics across all columns
df.describe()

,id,host_id,latitude,longitude,price,minimum_nights,number_of_reviews,reviews_per_month,calculated_host_listings_count,availability_365
count,4.889500e+04,4.889500e+04,48895.000000,48895.000000,48895.000000,48895.000000,48895.000000,38843.000000,48895.000000,48895.000000
mean,1.901714e+07,6.762001e+07,40.728949,-73.952170,152.720687,7.029962,23.274466,1.373221,7.143982,112.781327
std,1.098311e+07,7.861097e+07,0.054530,0.046157,240.154170,20.510550,44.550582,1.680442,32.952519,131.622289
min,2.539000e+03,2.438000e+03,40.499790,-74.244420,0.000000,1.000000,0.000000,0.010000,1.000000,0.000000
25%,9.471945e+06,7.822033e+06,40.690100,-73.983070,69.000000,1.000000,1.000000,0.190000,1.000000,0.000000
50%,1.967728e+07,3.079382e+07,40.723070,-73.955680,106.000000,3.000000,5.000000,0.720000,1.000000,45.000000
75%,2.915218e+07,1.074344e+08,40.763115,-73.936275,175.000000,5.000000,24.000000,2.020000,2.000000,227.000000
max,3.648724e+07,2.743213e+08,40.913060,-73.712990,10000.000000,1250.000000,629.000000,58.500000,327.000000,365.000000


In [7]:
# check for duplicated rows
df.duplicated().sum()

0

In [8]:
# check for null values across all columns
df.isna().sum()

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [9]:
# percentage of null values per column

for col in df.columns:
    col_null = df[col].isna().sum()
    print(f'{col} has {(col_null/len(df))*100:.3f}% null values')

id has 0.000% null values
name has 0.033% null values
host_id has 0.000% null values
host_name has 0.043% null values
neighbourhood_group has 0.000% null values
neighbourhood has 0.000% null values
latitude has 0.000% null values
longitude has 0.000% null values
room_type has 0.000% null values
price has 0.000% null values
minimum_nights has 0.000% null values
number_of_reviews has 0.000% null values
last_review has 20.558% null values
reviews_per_month has 20.558% null values
calculated_host_listings_count has 0.000% null values
availability_365 has 0.000% null values


# Summary

After the initial inspection, these are our findings;

1. We have 48k rows and 16 columns
2. All columns in the dataset are correctly formatted
3. There are no duplicate rows
4. We have null values in name, host_name, last_review, and review_per_month columns

=====================================================================================================

# 2. Column by Column Exploration

In this section, we will go through all the columns in this dataset **individually**. 

=====================================================================================================

## 2.1. Id Column

### Expectation?
The id column is the unique identifer for listing. 
We expect that the column should have distinct value -- No duplicates.

Also, we expect no null values

In [10]:
# Inspect the Id column

print(f'{df.columns[0]} has {df[df.columns[0]].isna().sum()} null values')
print(f'{df.columns[0]} has {df[df.columns[0]].duplicated().sum()} duplicate rows')
print(f'{df.columns[0]}: {df[df.columns[0]].dtype}')

id has 0 null values
id has 0 duplicate rows
id: int64


## Id Column Cleaning

No Null Values. No Duplicates. Correct Data Type.

Note: If duplicates had been found in the column, it would be investigated before dropping. duplicate Id could be data entry error amongst others.

## 2.2. Name Column

### Expectation?

The name column is the listing name for each listing. It should be a text based column.

In section 1, we saw that there is a 0.033% of null values in the columnns.

In [11]:
# Inspect the name column

print(f'{df.columns[1]} has {df[df.columns[1]].isna().sum()} null values')
print(f'{df.columns[1]} has {df[df.columns[1]].duplicated().sum()} duplicate rows')
print(f'{df.columns[1]}: {df[df.columns[1]].dtype}')

name has 16 null values
name has 989 duplicate rows
name: object


In [12]:
# filter out the null values as a dataframe

name_null_mask = df['name'].isna()
df.loc[name_null_mask]

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
2854,1615764,NaN,6676776,Peter,Manhattan,Battery Park City,40.71239,-74.01620,Entire home/apt,400,1000,0,NaN,NaN,1,362
3703,2232600,NaN,11395220,Anna,Manhattan,East Village,40.73215,-73.98821,Entire home/apt,200,1,28,2015-06-08,0.45,1,341
5775,4209595,NaN,20700823,Jesse,Manhattan,Greenwich Village,40.73473,-73.99244,Entire home/apt,225,1,1,2015-01-01,0.02,1,0
5975,4370230,NaN,22686810,Michaël,Manhattan,Nolita,40.72046,-73.99550,Entire home/apt,215,7,5,2016-01-02,0.09,1,0
6269,4581788,NaN,21600904,Lucie,Brooklyn,Williamsburg,40.71370,-73.94378,Private room,150,1,0,NaN,NaN,1,0
6567,4756856,NaN,1832442,Carolina,Brooklyn,Bushwick,40.70046,-73.92825,Private room,70,1,0,NaN,NaN,1,0
6605,4774658,NaN,24625694,Josh,Manhattan,Washington Heights,40.85198,-73.93108,Private room,40,1,0,NaN,NaN,1,0
8841,6782407,NaN,31147528,Huei-Yin,Brooklyn,Williamsburg,40.71354,-73.93882,Private room,45,1,0,NaN,NaN,1,0
11963,9325951,NaN,33377685,Jonathan,Manhattan,Hell's Kitchen,40.76436,-73.98573,Entire home/apt,190,4,1,2016-01-05,0.02,1,0
12824,9787590,NaN,50448556,Miguel,Manhattan,Harlem,40.80316,-73.95189,Entire home/apt,300,5,0,NaN,NaN,5,0


In [13]:
# check for stale listing

df.loc[name_null_mask, ['name', 'number_of_reviews', 'availability_365', 'price']].sort_values('number_of_reviews')

,name,number_of_reviews,availability_365,price
2854,NaN,0,362,400
6269,NaN,0,0,150
6567,NaN,0,0,70
6605,NaN,0,0,40
8841,NaN,0,0,45
12824,NaN,0,0,300
13059,NaN,0,0,67
15819,NaN,0,0,100
16071,NaN,0,365,130
18047,NaN,0,0,70


In [14]:
# filter out the stale listings

stale_mask = ((df['number_of_reviews'] == 0) 
              & (df['availability_365'] == 0)
             & df['name'].isna())
    
df[stale_mask].shape

(8, 16)

In [15]:
# drop the stale listings

df = df[~stale_mask].reset_index(drop=True)
df.shape

(48887, 16)

In [16]:
# fill the null values in name column with 'unknown name'

df['name'] = df['name'].fillna('Unknown')

In [17]:
# Validate the fill

df['name'].isna().sum()

0

In [18]:
# filter out the duplicates in the name column

dupes_name_mask = ((df['number_of_reviews'] == 0) &
                   (df.duplicated(subset=['name', 'host_id', 'neighbourhood', 
                                        'room_type'], keep=False))
)



In [19]:
# view filter

df.loc[dupes_name_mask, ['name', 'host_id', 'neighbourhood', 
                                        'room_type', 'number_of_reviews']].sort_values('name')

,name,host_id,neighbourhood,room_type,number_of_reviews
42768,1 BR Suite in Midtown Manhattan,24831061,Midtown,Private room,0
42809,1 BR Suite in Midtown Manhattan,24831061,Midtown,Private room,0
30694,1 bedroom apartment on Columbus Avenue,76104209,Upper West Side,Entire home/apt,0
31992,1 bedroom apartment on Columbus Avenue,76104209,Upper West Side,Entire home/apt,0
33159,20-25 minutes to downtown NYC/30-40 to midtown,89769986,Bedford-Stuyvesant,Private room,0
48199,"30 mins to Times Square!! 15 mins LGA, 25mins ...",260209224,Jackson Heights,Entire home/apt,0
44910,"30 mins to Times Square!! 15 mins LGA, 25mins ...",260209224,Jackson Heights,Entire home/apt,0
29533,307 east 44th street fully furnished,165898555,Midtown,Entire home/apt,0
29541,307 east 44th street fully furnished,165898555,Midtown,Entire home/apt,0
26079,A CLASSIC NYC NEIGHBORHOOD-EAST 86TH/5TH AVENUE,76104209,Upper East Side,Entire home/apt,0


In [20]:
# drop filter

df.sort_values('id', inplace=True)

df = df[
    ~dupes_name_mask |
    ~df.duplicated(subset=['name', 'host_id', 'neighbourhood', 'room_type'], keep='first')
].reset_index(drop=True)

In [21]:
print(f'{df.columns[1]} has {df[df.columns[1]].isna().sum()} null values')
print(f'{df.columns[1]} has {df[df.columns[1]].duplicated().sum()} duplicate rows')
print(f'{df.columns[1]}: {df[df.columns[1]].dtype}')
df.shape

name has 0 null values
name has 877 duplicate rows
name: object


(48783, 16)

## Name Column Cleaning

### Null Values
The `name` column had 16 missing values. Upon inspection, 8 of these listings had zero reviews and zero availability — indicating stale/abandoned listings with no transaction history and no active availability on the platform. These 8 rows were dropped.

The remaining 8 null listings had either reviews or availability, suggesting they were real active listings. Since the `name` column is descriptive and not used in analysis, nulls were filled with `'Unknown'` to preserve these rows.

### Duplicates
Duplicates were defined as rows sharing the same `name`, `host_id`, `neighbourhood`, and `room_type` — with `number_of_reviews == 0` on all occurrences. Listings with reviews were excluded from this definition as they represent real transaction history and are considered legitimate separate listings on the platform.

The `id` column was used as a tiebreaker — the lowest `id` (original listing) was kept and all subsequent duplicates were dropped.

**Duplicates before cleaning:** 989     
**Duplicates after cleaning:** 877     
**Duplicates dropped:** 112

**Rows before cleaning:** 48,858  
**Rows after cleaning:** 48,783  
**Rows dropped:** 75



## 2.3. Host_Id Column

### Expectation?
The `host_id` is a unique identifier given to each host on the platform. We can expect to have duplicate `host_id` as oppose to our `id` since a single host can decide to have multiple listings.      
From our basic inspection, we have seen that there are no null values and the data type is an `object`.

In [22]:
# Inspect the host_id column

print(f'{df.columns[2]} has {df[df.columns[2]].isna().sum()} null values')
print(f'{df.columns[2]} has {df[df.columns[2]].duplicated().sum()} duplicate rows')
print(f'{df.columns[2]}: {df[df.columns[1]].dtype}')

host_id has 0 null values
host_id has 11332 duplicate rows
host_id: object


In [23]:
# mapping host_id to host_name

host_mapping = df.groupby('host_id')['host_name'].nunique().sort_values(ascending=False)
host_mapping[host_mapping > 1]

Series([], Name: host_name, dtype: int64)

## Host_Id Column Cleaning

We expected that there would be duplicate host_id in the dataset because a host can have multiple listings. To confirm if the duplicates are not as a result of error, we have decided to map every `host_id` to a `host_name` to check if every id belongs to one name.
If the mapping had resulted to some `host_id` belonging to more than one name, more investigation would been done to why.

## 2.4. Host_Name Column

### Expectation?
`host_name` is the display name for hosts on the platform. Duplicates are expected and normal for this column since hosts can bear the same name and a single host can have multiple listings.   
In our basic inspection, we can expect to see 21 null values.

In [24]:
# Inspect the host_name column

print(f'{df.columns[3]} has {df[df.columns[3]].isna().sum()} null values')
print(f'{df.columns[3]} has {df[df.columns[3]].duplicated().sum()} duplicate rows')
print(f'{df.columns[3]}: {df[df.columns[3]].dtype}')

host_name has 21 null values
host_name has 37331 duplicate rows
host_name: object


In [25]:
# Filter out the null value from the main dataframe

host_name_mask = df['host_name'].isna()
df.loc[host_name_mask]

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
360,100184,Bienvenue,526653,NaN,Queens,Queens Village,40.72413,-73.76133,Private room,50,1,43,2019-07-08,0.45,1,88
2699,1449546,Cozy Studio in Flatbush,7779204,NaN,Brooklyn,Flatbush,40.64965,-73.96154,Entire home/apt,100,30,49,2017-01-02,0.69,1,342
5744,4183989,SPRING in the City!! Zen-Style Tranquil Bedroom,919218,NaN,Manhattan,Harlem,40.80606,-73.95061,Private room,86,3,34,2019-05-23,1.00,1,359
6074,4446862,Charming Room in Prospect Heights!,23077718,NaN,Brooklyn,Crown Heights,40.67512,-73.96146,Private room,50,1,0,NaN,NaN,1,0
6579,4763327,"Luxurious, best location, spa inc'l",24576978,NaN,Brooklyn,Greenpoint,40.72035,-73.95355,Entire home/apt,195,1,1,2015-10-20,0.02,1,0
8159,6292866,Modern Quiet Gem Near All,32722063,NaN,Brooklyn,East Flatbush,40.65263,-73.93215,Entire home/apt,85,2,182,2019-06-19,3.59,2,318
8253,6360224,"Sunny, Private room in Bushwick",33134899,NaN,Brooklyn,Bushwick,40.70146,-73.92792,Private room,37,1,1,2015-07-01,0.02,1,0
8847,6786181,R&S Modern Spacious Hideaway,32722063,NaN,Brooklyn,East Flatbush,40.64345,-73.93643,Entire home/apt,100,2,157,2019-06-19,3.18,2,342
9133,6992973,1 Bedroom in Prime Williamsburg,5162530,NaN,Brooklyn,Williamsburg,40.71838,-73.95630,Entire home/apt,145,1,0,NaN,NaN,1,0
9812,7556587,Sunny Room in Harlem,39608626,NaN,Manhattan,Harlem,40.82929,-73.94182,Private room,28,1,1,2015-08-01,0.02,1,0


In [26]:
# Check for stale listing

host_name_stale = ((df['host_name'].isna()) &
                  (df['number_of_reviews'] == 0) &
                  (df['availability_365'] == 0))
df.loc[host_name_stale]

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
6074,4446862,Charming Room in Prospect Heights!,23077718,NaN,Brooklyn,Crown Heights,40.67512,-73.96146,Private room,50,1,0,NaN,NaN,1,0
9133,6992973,1 Bedroom in Prime Williamsburg,5162530,NaN,Brooklyn,Williamsburg,40.71838,-73.95630,Entire home/apt,145,1,0,NaN,NaN,1,0
27729,21787094,Spacious and Bright Midtown East Apartment,415290,NaN,Manhattan,Upper East Side,40.76436,-73.96836,Entire home/apt,325,1,0,NaN,NaN,1,0
32140,25180572,Beautiful Cozy Studio Apt. UWS,119609345,NaN,Manhattan,Upper West Side,40.77701,-73.97667,Entire home/apt,180,4,0,NaN,NaN,1,0


In [27]:
# drop the stale listing

df = df[~host_name_stale].reset_index(drop=True)
df.shape

(48779, 16)

In [28]:
# map host_name to host_id from existing records and fill null host_names

host_name_to_id = df.dropna(subset=['host_name']).groupby('host_id')['host_name'].first()
df['host_name'] = df['host_name'].fillna(df['host_id'].map(host_name_to_id))

In [29]:
# check if we still have missing values

df['host_name'].isna().sum()

17

In [30]:
# Fill the null values in host_name witj 'unknown'

df['host_name'] = df['host_name'].fillna('unknown')

In [31]:
# validate the fill

df['host_name'].isna().sum()

0

In [32]:
# Validate the host_name column

print(f'{df.columns[3]} has {df[df.columns[3]].isna().sum()} null values')
print(f'{df.columns[3]} has {df[df.columns[3]].duplicated().sum()} duplicate rows')
print(f'{df.columns[3]}: {df[df.columns[3]].dtype}')

host_name has 0 null values
host_name has 37327 duplicate rows
host_name: object


## Host_Name Column Cleaning

After inspecting the null values in the `host_name` column, we realized that there we some stale listings. These were listings that do not have a `host_name`, no `number_of_reviews`, and no `availability_365`. We dropped the stale listings.    

We went on to try to map host_name to host_id from the existing records. This helps to check if there is a host_name for the same host_id in our existing record. Unfortunately, there wasn't.   
We decided to fill the null values in the `host_name` column with **unknown**


**Rows before cleaning:** 48,783   
**Rows after cleaning:** 48,779  
**Rows dropped:** 4

## 2.5. Neighbourhood_group Column

### Expectation
`neighbourhood_group` represents the borough in New York City where the listing is located.    
NYC has exactly 5 boroughs: Manhattan, Brooklyn, Queens, The Bronx, and Staten Island.    
Every listing should belong to one of these 5 values — no nulls, no unexpected categories.   
From our basic inspection, there are no null values, duplicates are expected.

In [33]:
# Inspect the neighbourhood_group column

print(f'{df.columns[4]} has {df[df.columns[4]].isna().sum()} null values')
print(f'{df.columns[4]} has {df[df.columns[4]].duplicated().sum()} duplicate rows')
print(f'{df.columns[4]}: {df[df.columns[4]].dtype}')

neighbourhood_group has 0 null values
neighbourhood_group has 48774 duplicate rows
neighbourhood_group: object


In [34]:
# validate if there are no unexpexted categories in the column

df['neighbourhood_group'].value_counts()

Manhattan        21593
Brooklyn         20076
Queens            5646
Bronx             1091
Staten Island      373
Name: neighbourhood_group, dtype: int64

## Neighbourhood_group Column Cleaning

No null values.   
No unexpected categories.  
The column is clean, no cleaning required.

## 2.6. Neighbourhood Column

### Expectation
`neighbourhood` represents the specific neighbourhood within each borough where the listing 
is located. Multiple listings can share the same neighbourhood which means duplicates are expected and 
normal.     
From basic inspection, no nulls expected. Values should be consistent with their corresponding `neighbourhood_group` (borough).

In [35]:
# Inspect the neighbourhood column

print(f'{df.columns[5]} has {df[df.columns[5]].isna().sum()} null values')
print(f'{df.columns[5]} has {df[df.columns[5]].duplicated().sum()} duplicate rows')
print(f'{df.columns[5]}: {df[df.columns[5]].dtype}')

neighbourhood has 0 null values
neighbourhood has 48558 duplicate rows
neighbourhood: object


In [36]:
# Check if neighbourhood values corresponds with their borough

df.groupby('neighbourhood')['neighbourhood_group'].nunique().sort_values(ascending=False).head(10)

neighbourhood
Allerton           1
Parkchester        1
New Dorp           1
New Dorp Beach     1
New Springville    1
NoHo               1
Nolita             1
North Riverdale    1
Norwood            1
Oakwood            1
Name: neighbourhood_group, dtype: int64

In [37]:
# check for nunmber of unique neighbourhood

df['neighbourhood'].nunique()

221

## Neighbourhood Column Cleaning

No null values.   
221 unique neighbourhood.   
Each neighbourhood exactly tied to one borough.   
No cleaning needed.

## 2.7. Latitude Column

### Expectation
`latitude` represents the geographic coordinate of each listing. No nulls expected. 
Duplicates are expected — different listings at different addresses will have unique coordinates, 
but listings in the same building could share coordinates.

NYC latitude range is approximately **40.4774 to 40.9176**. Any value outside this range 
is invalid and should be investigated.

In [38]:
# Inspect latitude column

print(f'{df.columns[6]} has {df[df.columns[6]].isna().sum()} null values')
print(f'{df.columns[6]} has {df[df.columns[6]].duplicated().sum()} duplicate rows')
print(f'{df.columns[6]}: {df[df.columns[6]].dtype}')

latitude has 0 null values
latitude has 29737 duplicate rows
latitude: float64


In [39]:
# Check for latittude range

print(f'{df["latitude"].min()}, {df["latitude"].max()}')

40.49979, 40.91306


## Latitude Column Cleaning

No null values.   
Falls within NYC bounds.      
No cleaning required.

## 2.8. Longitude Column

### Expectation

`longitude` represents the geographic coordinate of each listing. No nulls expected. 
Duplicates are expected — different listings at different addresses will have unique coordinates, 
but listings in the same building could share coordinates.
NYC longitude range is approximately **-74.2591 to -73.7004**. Any value outside this range 
is invalid and should be investigated.

In [40]:
# Inspect longitude column

print(f'{df.columns[7]} has {df[df.columns[7]].isna().sum()} null values')
print(f'{df.columns[7]} has {df[df.columns[7]].duplicated().sum()} duplicate rows')
print(f'{df.columns[7]}: {df[df.columns[7]].dtype}')

longitude has 0 null values
longitude has 34077 duplicate rows
longitude: float64


In [41]:
# Check for longitude range

print(f'{df["longitude"].min()}, {df["longitude"].max()}')

-74.24442, -73.71299


## Longitude Column Cleaning

No null values.   
Falls within NYC bounds.      
No cleaning required.

## 2.9. Room_type Column

### Expectation
`room_type` represents the type of room the listing is.    
air_bnb has exactly 3 `room_type`: Entire Home/apt, Private Room, and Shared Room.   
Every listing should belong to one of these 3 `room_type` — no nulls, no unexpected categories.   
From our basic inspection, there are no null values, duplicates are expected.

In [42]:
# Inspect room_type column

print(f'{df.columns[8]} has {df[df.columns[8]].isna().sum()} null values')
print(f'{df.columns[8]} has {df[df.columns[8]].duplicated().sum()} duplicate rows')
print(f'{df.columns[8]}: {df[df.columns[8]].dtype}')

room_type has 0 null values
room_type has 48776 duplicate rows
room_type: object


In [43]:
# check for unexpected categories

df['room_type'].value_counts()

Entire home/apt    25351
Private room       22271
Shared room         1157
Name: room_type, dtype: int64

## Room_type Column Cleaning

There are no null values.    
There are no unexpected categories.
No cleaning required.